### ISSUE TITLE
option to keep table header visible when data length is 0

### ISSUE THREAD

**Issue by frantic1048** [buefy/buefy#568]
### Overview of the problem

**Buefy** version: 0.6.2
**Vuejs** version: 2.5.2
**OS/Browser**: various

### Description

This issue is a feature request:

- an option for b-table, that keeps table being rendered with the table header
- or, make table's empty slot replace only the table body area
- or somewhat equivalent

### Context

I implemented some filter controls within custom table header. The table data is being filtered according to filter controls status in the table header.

When I set [a combination of] filter options that produces no data as a result. The table was entirely replaced by its `empty` slot. I could only refresh the page or go backward(if the filter state is bound on URL query string), to make the filter controls in table header back. Then I can try other filter options.

If the table header could be rendered when data is empty, thus I could still interact with filter controls on table header, without refreshing or going backward.
 
---

update

2018-01-31: fix grammar in the text

**Comment by grebebird** [comment]
+1 for either fixing this issue, or updating Buefy's documentation. Oh please please do one or the other to save future developers' time. :)

The saddest thing about this issue is that Buefy's own example of this ([Sandbox with custom template](https://buefy.github.io/#/documentation/table)) is flawed, which led our team to a bunch of time spent working in circles to try to recreate what we had been seeing in the documentation. 

I CodePen'ed Buefy's documentation with the "isEmpty" flag set to TRUE **_first_** to demonstrate this issue. The headers don't show up, but if you toggle the "isEmpty" flag to FALSE and **_then_** TRUE, the headers are back, in the way we were expecting the first time through.

https://codepen.io/anon/pen/NYExzV?&editors=101

**Comment by ivalkenburg** [comment]
This happens because the newColums property isn't being populated when there is no data. No data means there are no table column component created that register itself to the parent table component which populates the newColumns property.

**Comment by willhoyle** [comment]
I'll look into the src tonight and see if it can be fixed but for now you can kind of do the same hack as the Documentation:

```js
<template>
    <b-table :data="fake ? fakeData : data">
    </b-table>
</template>

<script>
export default {
data() {
 return {
    fake: true,
    data: [],
    fakeData: [{id: 1, name: " "}] // fill in some fake empty values for your columns
},
  mounted() {
    this.$nextTick(() => {
      this.fake = false
    })
  }
}
</script>
```

It's far from ideal and extremely hacky but in a pinch this can work.

**Comment by willhoyle** [comment]
Okay had time to take a proper look at the source.

Unfortunately, this can't be easily fixed.

Summary of the inner workings of the component:

The main b-table component has a data property called newColumns. When the data( ) property is first initialized, it is seeded with either an empty array as default, or with the props.columns supplied by the user.

When you don't supply a props.columns,  this.newColumns = [ ]. 

The <thead> tag only renders when this.newColumns has items. So how does it get populated? Only once b-table starts rendering rows (in tbody, and in a v-for loop on the <tr> tag). 

When b-table starts displaying rows, the b-table-columns from the default slot start rendering. When a b-table-columns component's beforeMount hook is triggered, it accesses this.$parent.newColumns and adds itself to the list of columns (it checks to make sure it's not already there first by using the label or props.customKey). 

When the props.data = [ ] upon first render, the b-table component can't try to render rows so none of the b-table-columns can register themselves to the this.newColumns array.

There's a simple solution though. Just supply the columns prop to b-table with the headers you need. Just make sure that you either give the same labels as the ones in b-table-column, or the same customKey if you're using those.

TLDR; if you don't care about the inner workings of the lib. look at this code from the documentation page. I added the column prop and everything works as intended.

```vue
<template>
  <section>
    <b-field grouped group-multiline>
      <div class="control">
        <b-switch v-model="isBordered">Bordered</b-switch>
      </div>
      <div class="control">
        <b-switch v-model="isStriped">Striped</b-switch>
      </div>
      <div class="control">
        <b-switch v-model="isNarrowed">Narrowed</b-switch>
      </div>
      <div class="control">
        <b-switch v-model="isHoverable">Hoverable</b-switch>
      </div>
      <div class="control">
        <b-switch v-model="isFocusable">Focusable</b-switch>
      </div>
      <div class="control">
        <b-switch v-model="isLoading">Loading state</b-switch>
      </div>
      <div class="control">
        <b-switch v-model="isEmpty">Empty</b-switch>
      </div>
      <div class="control">
        <b-switch v-model="hasMobileCards">
          Mobile cards
          <small>(collapsed rows)</small>
        </b-switch>
      </div>
    </b-field>

    <b-table
      :data="isEmpty ? [] : data"
      :columns="columns"
      :bordered="isBordered"
      :striped="isStriped"
      :narrowed="isNarrowed"
      :hoverable="isHoverable"
      :loading="isLoading"
      :focusable="isFocusable"
      :mobile-cards="hasMobileCards"
    >
      <template slot-scope="props">
        <b-table-column field="id" label="ID" width="40" numeric>{{ props.row.id }}</b-table-column>

        <b-table-column field="first_name" label="First Name">{{ props.row.first_name }}</b-table-column>

        <b-table-column field="last_name" label="Last Name">{{ props.row.last_name }}</b-table-column>

        <b-table-column field="date" label="Date" centered>
          <span class="tag is-success">{{ new Date(props.row.date).toLocaleDateString() }}</span>
        </b-table-column>

        <b-table-column label="Gender">
          <b-icon pack="fas" :icon="props.row.gender === 'Male' ? 'mars' : 'venus'"></b-icon>
          {{ props.row.gender }}
        </b-table-column>
      </template>

      <template slot="empty">
        <section class="section">
          <div class="content has-text-grey has-text-centered">
            <p>
              <b-icon icon="emoticon-sad" size="is-large"></b-icon>
            </p>
            <p>Nothing here.</p>
          </div>
        </section>
      </template>
    </b-table>
  </section>
</template>

<script>
export default {
  data() {
    const data = [
      {
        id: 1,
        first_name: "Jesse",
        last_name: "Simmons",
        date: "2016/10/15 13:43:27",
        gender: "Male"
      },
      {
        id: 2,
        first_name: "John",
        last_name: "Jacobs",
        date: "2016/12/15 06:00:53",
        gender: "Male"
      },
      {
        id: 3,
        first_name: "Tina",
        last_name: "Gilbert",
        date: "2016/04/26 06:26:28",
        gender: "Female"
      },
      {
        id: 4,
        first_name: "Clarence",
        last_name: "Flores",
        date: "2016/04/10 10:28:46",
        gender: "Male"
      },
      {
        id: 5,
        first_name: "Anne",
        last_name: "Lee",
        date: "2016/12/06 14:38:38",
        gender: "Female"
      }
    ];
    let columns = [
      { label: "ID" },
      { label: "First Name" },
      { label: "Last Name" },
      { label: "Date" },
      { label: "Gender" }
    ];
    return {
      data,
      columns,
      isEmpty: true,
      isBordered: false,
      isStriped: false,
      isNarrowed: false,
      isHoverable: false,
      isFocusable: false,
      isLoading: false,
      hasMobileCards: true
    };
  }
};
</script>
```

I hope this can help anyone that was stuck.

**Comment by willhoyle** [comment]
That’s great. If you need any help for anything I wouldn’t mind being involved. 

Turns out my post above doesn’t really fix it completely and brought out other issues.

**Comment by jtommy** [comment]
https://github.com/buefy/buefy/issues/1691

**Comment by s-nesbigall** [comment]
Hi,
is there any chance to revive this issue?
The current behaviour is not consistent:

If the table gets loaded without data -> headers are not visible
If the data changes during runtime to an empty array -> headers are visible

For now I try the workaround described above by @willhoyle and provide additional columns by prop so they show up, if the data array is empty.

Best regards

**Comment by jtommy** [comment]
Sure, follow the other issue

**Comment by florentyu** [comment]
Hello,

@willhoyle workaround doesn't work in [version 0.8.10](https://github.com/buefy/buefy/releases/tag/v0.8.10). In this version, columns in `columns` attribute are rendered.
Moreover in this case, `sortable` attribute doesn't work in the `props` slot.
